### PR TITLE
Revamped Ignite Documentation

### PR DESCRIPTION
**Pull request by jamonholmgren**
This long-running PR will hold our new documentation efforts until we merge them in.

A few notes:

* We should add at least one screenshot of each component into each example doc, so people know what they look like
* It's important to show a fairly comprehensive example for each property, and I've tried to do this here. We shouldn't skimp out on information
* @yulolimum's PR #1965 or future PRs might change some of these components, so let's be careful not to go too far until we've coordinated with him

cc @curlyCoder15 @lindboe

### CODE REVIEW

File: docs/Components-Wallpaper.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Wallpaper Component
+
+[Back to all components](./Components.md)
+
+
+Ignite's `Wallpaper` component renders an image background to ypur component or a screen. It is composed of an [Image](https://reactnative.dev/docs/image) component. You can also create your own presets for the background image.
```
Comment (jamonholmgren) [inline docs/Components-Wallpaper.md:6]: s/ypur/your

File: docs/Components-Wallpaper.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Props
+
+### `style`
+
+The `style` prop is optional, it overrides the default style and is used for padding and margin. It is of the type `StyleProp<ImageStyle>`. See React Native docs on [ImageStyle](https://reactnative.dev/docs/image#style) for more information.
+
+```tsx
+<Wallpaper
+  style={{ margunHorizontal: 5, paddingVertical:10 }}
```
Comment (jamonholmgren) [inline docs/Components-Wallpaper.md:24]: "margun"

File: docs/Components-Wallpaper.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Presets allow you to have a consistent look and feel across your app without having to redefine the styles all the time.
+
+```tsx
+<Wallpaper preset="stretch" />
+```
+
+You'll want to customize these presets and add more. You can do this in the `wallpaper.presets.ts` file.
+
+If you find yourself overriding styles with the `style` prop a lot, you probably need a new preset for that use case.
```
Comment (jamonholmgren) [inline docs/Components-Wallpaper.md:48]: We should list all available presets. In this case, it's just *stretch*, which we can note.

File: docs/Components-Wallpaper.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+### Custom Presets
+
+Feel free to add your own presets by emulating the style you see with the provided presets. In a typical Infinite Red project, we will have a dozen or more presets that we use across the project.
```
Comment (jamonholmgren) [inline docs/Components-Wallpaper.md:52]: For the Wallpaper component this isn't necessarily the case. Refine this part to reflect what we actually do on an Ignite IR project.

File: docs/Components.md [hunk 5]
```diff
@@ -150,3 +150,17 @@ This is an enhanced version of the built-in React Native Text component. It adds
...
+
+This component renders an image background to a component or a screen. 
+
+```tsx
+ <Wallpaper
+ style={{ 
+    marginHorizontal: 10,
+    paddingVertical: 10,
+  }} 
+ backgroundImage={$newImage}
```
Comment (jamonholmgren) [inline docs/Components.md:162]: This shouldn't have a $ on it, as that's reserved for styles.

Indentation isn't right on this, too.

**Review by jamonholmgren — COMMENTED** [review]
Some changes @curlyCoder15

File: docs/Components-Wallpaper.md (hunk 4 above)
Comment (curlyCoder15) [reply docs/Components-Wallpaper.md:52]: makes sense

File: docs/Components-TextField.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  preset="default"
+/>
+
+```
+
+## Props
+
+### `label`
+
+The `label` prop is optional. We encourage you to not use this but rather use the `labelTx` prop instead.
```
Comment (silasjmatson) [inline docs/Components-TextField.md:29]: Is this true? In the case that we don't have a translation available for the user's locale, wouldn't we want the `label` regardless? I'd be curious what other members of the team think about this, I don't have much experience with how we do things.

I'd also add a note that if `label` is missing `labelTx` must be set for any label text to be present.

File: docs/Components-TextField.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+<TextField
+  value={input}
+  onChangeText={(value) => setInput(value)}
+  label="Name"
+/>
+```
+
+### `labelTx`
+
+The `labelTx` prop is optional. The label text is looked up by [i18n](https://www.i18next.com/) (used for internationalization)
```
Comment (silasjmatson) [inline docs/Components-TextField.md:41]: I think the library we use is `i18n-js`, so I'd probably link to that, and maybe link to the `Component-Text.md` documentation, as these props are passed to that component.

File: docs/Components-TextField.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+<TextField
+  value={input}
+  onChangeText={(value) => setInput(value)}
+  placeholderTx="signup.name"
+/>
+```
+
+### `style`
+
+The `style` prop is optional. It overrides container style useful for margins & padding.
```
Comment (silasjmatson) [inline docs/Components-TextField.md:77]:
```suggestion
The `style` prop is optional. It overrides container style and is useful for margins & padding.
```

File: docs/Components-TextField.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+The `inputStyle` prop is optional. It overrides the input style.
+
+```tsx
+<TextField
+  value={input}
+  onChangeText={(value) => setInput(value)}
+  labelTx="signup.name"
+  placeholderTx="signup.name"
+  style={{ paddingHorizontal:10 }}
```
Comment (silasjmatson) [inline docs/Components-TextField.md:99]: I would remove the `style` prop here. In the specific prop examples, I think we only want required props plus the prop in the section. Basically the smallest valid use case that demonstrates the prop is what we're looking for.

File: docs/Components-TextField.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+The `preset` prop is optional. Use preset to setup various look and feels.
+
+```tsx
+<TextField
+  value={input}
+  onChangeText={(value) => setInput(value)}
+  labelTx="signup.name"
+  placeholderTx="signup.name"
+  style={{ paddingHorizontal:10 }}
+  inputStyle={{ color:color.palette.black }}
```
Comment (silasjmatson) [inline docs/Components-TextField.md:115]: As above, remove props not needed for the example -- `style` and `inputStyle` here.

File: docs/Components-TextField.md (hunk 6 above)
Comment (curlyCoder15) [reply docs/Components-TextField.md:29]: The label prop is optional and tbh I prefer that as I sometimes don't use labels for text field(personal preference) - eg login forms.

File: docs/Components.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+```
+
+[Full Button Component Documentation](./Components-Button.md)
+
+### Checkbox
+
+This is a component that renders a [`TouchableOpacity`](https://reactnative.dev/docs/touchableopacity) with a view to hold a checkbox and a text label.
+
+```tsx
+[on, setOn] = useState(false);
```
Comment (jamonholmgren) [inline docs/Components.md:64]:
```suggestion
const [on, setOn] = useState(false);
```

File: docs/Components.md [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+  }}
+  onToggle={setOn}
+/>
+```
+
+[Full Checkbox Component Documentation](./Components-Checkbox.md)
+
+### FormRow
+
+This is a component that renders a horizontal container for a form row.
```
Comment (jamonholmgren) [inline docs/Components.md:95]: I think this needs a little more explanation here. The example should probably show some inputs as well.

File: docs/Components.md [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+[Full FormRow Component Documentation](./Components-FormRow.md)
+
+### GradientBackground
+
+This is a convenience component that wraps [`LinearGradient`]((https://docs.expo.dev/versions/latest/sdk/linear-gradient/)) from `expo-linear-gradient` and handles the styling for you.
+
+```tsx
+<GradientBackground colors={["#422443", "#281b34"]} />
+```
```
Comment (jamonholmgren) [inline docs/Components.md:117]: Missing link to full docs

File: docs/Components.md [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+ <Wallpaper
+  style={{
+    marginHorizontal: 10,
+    paddingVertical: 10,
+  }}
+  backgroundImage={newImage}
+  preset="stretch"
+ />
+```
+[Full Text Component Documentation](./Components-Wallpaper.md)
```
Comment (jamonholmgren) [inline docs/Components.md:235]: Let's add a section at the bottom that talks about custom components.

Mention the component generator:

```
npx ignite-cli generate component MyCustomButton
```

File: docs/Components-AutoImage.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
+# Text Component
```
Comment (jamonholmgren) [inline docs/Components-AutoImage.md:1]: Should be AutoImage

File: docs/Components-AutoImage.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+[Back to all components](./Components.md)
+
+Ignite's `AutoImage` Component is an enhanced version of the built-in React Native [Image](https://reactnative.dev/docs/image) component. It automatically resizes the image view to fit the size of the image and ensures you don't need to explicitly set the image size props on every image. You can use it in place of the built-in `Image` component. We recommend importing it `as Image`, as there are no downsides to doing so, and ensures images will always be resized if needed.
```
Comment (jamonholmgren) [inline docs/Components-AutoImage.md:5]:
```suggestion
Ignite's `AutoImage` Component is an enhanced version of the built-in React Native [Image](https://reactnative.dev/docs/image) component. It automatically resizes the image view to fit the size of the image and ensures you don't need to explicitly set the image size props on every image. You can use it in place of the built-in `Image` component. We at Infinite Red often import it `as Image`, as there are no significant downsides to doing so, and ensures images will always be resized if needed.
```

File: docs/Components-Checkbox.md [hunk 17]
```diff
@@ -0,0 +1,150 @@
+# Checkbox Component
+
+[Back to all components](./Components.md)
+
+The `Checkbox` component is a simple checkbox that can be used to toggle a boolean value. It is composed of a [`TouchableOpacity`](https://reactnative.dev/docs/touchableopacity) with a view to hold a checkbox and a text label.
+
+```tsx
+[on, setOn] = useState(false);
```
Comment (jamonholmgren) [inline docs/Components-Checkbox.md:8]: I won't mark the rest of these, but make sure we get `const` on there.

```suggestion
const [on, setOn] = useState(false);
```

File: docs/Components-Header.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+```tsx
+  <Header
+    leftIcon="back"
+    onLeftPress={() => navigation.goBack()}
+  />
+```
+
+### `rightIcon`
+
+The `rightIcon` is an optional prop that is used to set the icon for the right navigation button. Options are 'bullet', and 'bug'. Custom icons can be created by using the [`Icon` component](./Components-Icon.md#custom-icons).
```
Comment (jamonholmgren) [inline docs/Components-Header.md:59]: Can you explain a bit more how to create a custom icon? Like, do you just create an icon preset and then put the string name in here? Or do you have to provide the <Icon /> JSX directly?

File: docs/Guide.md [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+- React Native
+- React Navigation 6
+- MobX-State-Tree [(Why not Redux?)](https://github.com/infinitered/ignite/blob/master/docs/MobX-State-Tree.md)
+- MobX-React-Lite
+- TypeScript
+- AsyncStorage (integrated with MST for restoring state)
+- apisauce (to talk to REST servers)
+- Flipper-ready
+- Reactotron-ready (and pre-integrated with MST)
+- Supports Expo (and Expo web) out of the box
```
Comment (jamonholmgren) [inline docs/Guide.md:32]:
```suggestion
- Supports Expo (and Expo web) out of the box
- About a dozen prebuilt [components](./Components.md) to build out your UI with
```

File: docs/Guide.md [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+Learn more in our [Navigation](./Navigation.md) documentation.
+
+### Components
+
+Ignite comes with some prebuilt, flexible, and customizable components. Unlike most component libraries, it's not built to drop in out of the box, but rather with custom design in mind (you do have a nice, custom design, don't you?)
+
+Ignite works fine with other component libraries, but the built-in component system works quite well for most apps.
+
+Check out the [Components][./components.md] documentation.
```
Comment (jamonholmgren) [inline docs/Guide.md:56]:
```suggestion
Check out the [Components](./components.md) documentation.
```

File: docs/Guide.md [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+
+You can learn more about why we chose these tests and how to use them in the [Testing](./Testing.md) docs.
+
+### Styling
+
+Ignite's approach to styling is, like many other things in Ignite, straightforward and blunt.
+
+We don't use `StyleSheet.create()` as a general rule, as it doesn't provide any real benefits over bare objects.
+
+We instead use a strategy of constants, colocated with our components, prefixed with `$`, and typed with TypeScript:
```
Comment (jamonholmgren) [inline docs/Guide.md:70]:
```suggestion
We instead use a strategy of constants, colocated with our components, all upper-case, and typed with TypeScript:
```

File: docs/Guide.md [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+
+```tsx
+import { View, ViewStyle } from "react-native"
+import { palette } from "../theme"
+
+const $container: ViewStyle = {
```
Comment (jamonholmgren) [inline docs/Guide.md:76]:
```suggestion
const CONTAINER: ViewStyle = {
```

File: docs/Guide.md [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+  flex: 1,
+  backgroundColor: palette.bgColor,
+}
+
+const MyComponent = () => {
+  return <View style={$container}>...</View>
```
Comment (jamonholmgren) [inline docs/Guide.md:82]:
```suggestion
  return <View style={CONTAINER}>...</View>
```

**Comment by jamonholmgren** [comment]
I am going to merge this, and future docs updates will happen over on the [Maverick](https://github.com/infinitered/ignite/milestone/19) branch/milestone. We won't promote the docs until we have the full set, released with Maverick (v8).

**Pull request merged**